lib/src/https_openssl.rs

```diff
@@ -1121,32 +1106,129 @@ pub struct Listener {
     active: bool,
 }
 
+impl CertificateResolver for Listener {
+    type Error = ListenerError;
+
+    fn get_certificate(
+        &self,
+        fingerprint: &CertificateFingerprint,
+    ) -> Option<ParsedCertificateAndKey> {
+        let resolver = self
+            .resolver
+            .lock()
+            .map_err(|err| ListenerError::LockError(err.to_string()))
+            .ok()?;
+
+        resolver.get_certificate(fingerprint)
+    }
+
+    fn add_certificate(
+        &mut self,
+        opts: &sozu_command_lib::proxy::AddCertificate,
+    ) -> Result<CertificateFingerprint, Self::Error> {
+        let (fingerprint, certificate_and_key) = {
+            let mut resolver = self
+                .resolver
+                .lock()
+                .map_err(|err| ListenerError::LockError(err.to_string()))?;
+
+            let fingerprint = resolver
+                .add_certificate(opts)
+                .map_err(|err| ListenerError::ResolverError(err))?;
+
+            let certificate_and_key = resolver
+                .get_certificate(&fingerprint)
+                .expect("to retrieve certificate and key has they just been inserted");
```

Comment:
Why the `expect`, why not return a `ListenerError` ?
I don't understand the error message here. Please clarify.